MerkleSumTree.md

```diff
@@ -0,0 +1,213 @@
+
+### Digests
+
+The digest for any node in the SMST is calculated in partially the same manner as the regular SMT. The main differences are that the sum is included in the digest preimage - meaning the hash of any node's data includes its sum. But also the sum is appended to the digest of the node, resulting in a digest that looks like:
+
+`digest = [node digest]+[8 byte hex sum]`
+
+Therefore for the following node types the digests are as follows:
+
+- Inner Nodes
+  - Prefixed `[]byte{1}`
+  - `digest = hash([]byte{1} + leftChild.digest + rightChild.digest + hex(leftChildSum+rightChildSum))+[8 byte hex sum]`
+- Extension Nodes
+  - Prefixed `[]byte{2}`
+  - `digest = hash([]byte{2} + pathBounds + path + child.digest + hex(childSum))+[8 byte hex sum]`
+- Sum Leaf Nodes
+  - Prefixed `[]byte{0}`
+  - `digest = hash([]byte{0} + path + value + hexSum)+[8 byte hex sum]`
+- Lazy Nodes
+  - Prefix of the actual node type is stored in the digest
+  - `digest = persistedDigest`
+
+This means that with a hasher such as `sha256.New()` whose hash size is 32 bytes, the digest of any node will be 40 bytes in length.
+
+### Sum Leaves
+
+The SMST introduces a new node type, the `sumLeafNode` which is almost identical to a `leafNode` from the SMT. However, it includes a `sum` field which is a `[8]byte` hexadecimal representation of the `uint64` sum of the node. In an SMST the `sumLeafNode` replaces the `leafNode` type.
+
+### Visualisations
+
+The following diagrams are representations of how the tree and its components can be visualised.
+
+#### General Tree Structure
+
+In reality the only nodes that actually hold a sum value are the `sumLeafNode` types and all other nodes store their sum as part of their digest. However, for the sake of visualisation the sum is included in all nodes.
+
+```mermaid
+graph TB
+	subgraph Root
+		A1["Digest: Hash(Hash(Path+Hash1)+Hash(Hash2+(Hash(Hash3+Hash4)))+Hex(20))+Hex(20)"]
+        A2[Sum: 20]
+	end
+	subgraph BI[Inner Node]
+		B1["Digest: Hash(Hash2+(Hash(Hash3+Hash4))+Hex(12))+Hex(12)"]
+        B2[Sum: 12]
+	end
+	subgraph BE[Extension Node]
+		B3["Digest: Hash(Path+Hash1+Hex(8))+Hex(8)"]
+        B4[Sum: 8]
+	end
+	subgraph CI[Inner Node]
+		C1["Digest: Hash(Hash3+Hash4+Hex(7))+Hex(7)"]
+        C2[Sum: 7]
+	end
+	subgraph CL[Sum Leaf Node]
+		C3[Hash2]
+        C4[Sum: 5]
+	end
+	subgraph DL1[Sum Leaf Node]
+		D1[Hash3]
+        D2[Sum: 4]
+	end
+	subgraph DL2[Sum Leaf Node]
+		D3[Hash4]
+        D4[Sum: 3]
+	end
+	subgraph EL[Sum Leaf Node]
+		E1[Hash1]
+        E2[Sum: 8]
+	end
+	Root-->|0| BE
+	Root-->|1| BI
+	BI-->|0| CL
+	BI-->|1| CI
+	CI-->|0| DL1
+	CI-->|1| DL2
+	BE-->EL
+```
+
+#### Sum Hex Digests
+
+The following diagram shows the structure of the digests of the nodes within the tree in a simplified manner, again only the `sumLeafNode` objects have a `sum` field but for visualisation purposes the sum is included in all nodes.
+
+```mermaid
+graph TB
```

Comment:
Inner nodes dont need the path - they represent simply a branch in the tree. Adding a path would be pointless IMO. It seems hard to understand because the extension nodes hold a path (ie the common prefix with the leaf node). In reality we should maintain that only the leaf nodes (and by extension 🥁 the extension nodes as well) hold any path. Otherwise we add unnecessary bloat to the tree.